webapp/src/immersivereader.tsx

```diff
@@ -0,0 +1,126 @@
+import * as data from "./data";
+import * as auth from "./auth";
+import * as core from "./core";
+import * as sui from "./sui";
+import * as ImmersiveReader from '@microsoft/immersive-reader-sdk';
+
+import Cloud = pxt.Cloud;
+
+export type ImmersiveReaderToken = {
+    token: string;
+    subdomain: string;
+    expiration: number;
+    expiresIn: number;
+}
+
+interface ImmersiveReaderProps {
+    content: string;
+}
+
+
+function beautifyText(content: string): string {
+    const cleaningFuncs = [cleanImages, cleanBlockAnnotation, cleanBold];
+    let contentWIP = content;
+
+    cleaningFuncs.forEach(clean => {
+        contentWIP = clean(contentWIP);
+    })
+
+    return contentWIP;
+
+    // Change ``|| around blocks to {}
+    function cleanBlockAnnotation(content: string): string {
+        const blockLiteral = /``\|\|([\w|\s]+:[\w|\s]+)\|\|``/
+        const blockRegex = new RegExp(blockLiteral, "g");
+        const blocksRemoved = content.replace(blockRegex, (blockDef) => {
+            const match = blockDef.match(blockLiteral);
+            return "{" + match[1] + "}";
+        })
+        return blocksRemoved;
+    }
+
+    // Don't show any images
+    function cleanImages(content: string): string {
+        const imageRegEx = /(!\[.*\]\(.*\))/g;
+        const imagesRemoved = content.replace(imageRegEx, "");
+        return imagesRemoved;
+    }
+
+    // Remove ** around bold text, replace with "
+    function cleanBold(content: string): string {
+        const boldLiteral = /\*\*([\w|\s]+)\*\*/
+        const boldRegex = new RegExp(boldLiteral, "g");
+        const boldRemoved = content.replace(boldRegex, (boldWord) => {
+            const match = boldWord.match(boldLiteral);
+            return "\"" + match[1] + "\""
+        })
+        return boldRemoved;
+    }
+}
+
+function getTokenAsync(): Promise<ImmersiveReaderToken> {
+    if (Cloud.isOnline()) {
+        const storedToken = pxt.storage.getLocal('immReader');
+        const cachedToken: ImmersiveReaderToken = storedToken ? JSON.parse(storedToken) : undefined;
+
+        if (!storedToken || (Date.now() / 1000 > cachedToken.expiration)) {
+            return auth.apiAsync("/api/immreader").then(res => {
+                if (res.statusCode == 200 ) {
+                    pxt.storage.setLocal('immReader', JSON.stringify(res.resp));
+                    return res.resp;
+                } else {
+                    pxt.storage.removeLocal("/api/immreader");
+                    console.log("immersive reader fetch token error: " + JSON.stringify(res.err));
+                    return Promise.reject(new Error("token"));
+                }
+            });
+        } else {
+            pxt.tickEvent("immersiveReader.cachedToken");
+            return Promise.resolve(cachedToken);
+        }
+    } else {
+        return Promise.reject(new Error("offline"));
+    }
+}
+
+export function launchImmersiveReader(content: string) {
+    pxt.tickEvent("immersiveReader.launch");
```

Comment:
hmm i wonder if it would be useful to know what specific content the immersive reader is being launched for? maybe not the entire content but the tutorial id? not sure... and i guess this could also be done with an "immersiveReader.button.click" handler on the tutorial.tsx end

Reply:
oh yeah I'll add the id!

Reply:
(is tutorial id the tutorial name, or the tutorial? or something else >>)
![image](https://user-images.githubusercontent.com/18712271/111673228-fef32c80-87d7-11eb-8dab-fbdd67e2162f.png)

Reply:
oh yeah, i think most of our ticks use tutorialOptions.tutorial... we're not entirely consistent >>

Reply:
excellentttt